### PR TITLE
BUG: Windows Compliance Update

### PR DESCRIPTION
**Pull request by rstoneback**
# Description

Addresses #790, partial #57

Working on adding windows compatibility.

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

online testing.

**Test Configuration**:
* Operating system: MacOS
* Version number: Python 3.7

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

If this is a release PR, replace the first item of the above checklist with the release
checklist on the wiki: https://github.com/pysat/pysat/wiki/Checklist-for-Release

### CODE REVIEW

**Comment by rstoneback** [comment]
Down to two types of errors. There is some time issue with netcdf4
```
left =                      altitude   latitude  ...        slt      uts
562
epoch                                     ...       ...4526  ...  19.546392  86200.0
563
2009-01-01 23:58:20     400.0  42.452482  ...  19.958763  86300.0
564

565
[864 rows x 6 columns]
566
right =                                altitude   latitude  ...        slt      uts
567
Epoch                                     ... 19.546392  86200.0
568
1969-12-31 23:59:57.852516352     400.0  42.452482  ...  19.958763  86300.0
569
```

and the calc_solar_local_time is different on windows for some reason.
```
        # This works because test instrument longitude ranges from 0-360 deg
678
>       assert (abs(self.py_inst['slt'].values
679
                    - self.py_inst['longitude'].values / 15.0)).max() < 1.0e-6
680
E       assert 23.999556885555556 < 1e-06
681
E        +  where 23.999556885555556 = <built-in method max of numpy.ndarray object at 0x0000020ADD25F390>()
682
E        +    where <built-in method max of numpy.ndarray object at 0x0000020ADD25F390> = array([23.99955689]).max
683
E        +      where array([23.99955689]) = abs((array([23.99955689]) - (array([0.]) / 15.0)))
684
E        +        where array([23.99955689]) = Epoch\n2009-01-01    23.999557\nName: slt, dtype: float64.values
685
E        +        and   array([0.]) = Epoch\n2009-01-01    0.0\nName: longitude, dtype: float64.values
686

```

**Comment by rstoneback** [comment]
@jklenzing FYI. I grabbed stuff from your branch and extended it a bit.

**Comment by jklenzing** [comment]
The conditions for the slt tests should be updated to use a modulus for 24 hours.  The difference at one point is 23.999556885555556.

**Comment by jklenzing** [comment]
The netcdf errors are coming from pandas in the final assert statements
```
ValueError: Can only compare identically-labeled Series objects
```

**Comment by rstoneback** [comment]
> The netcdf errors are coming from pandas in the final assert statements
> 
> ```
> ValueError: Can only compare identically-labeled Series objects
> ```

Yeah, that's coming up since the times are different. For some reason the files written to netcdf in windows are coming back with times in 1969.

**Comment by jklenzing** [comment]
> For some reason the files written to netcdf in windows are coming back with times in 1969.

🙀

**Comment by rstoneback** [comment]
> The conditions for the slt tests should be updated to use a modulus for 24 hours. The difference at one point is 23.999556885555556.

Modulus isn't going to help here, less than 24.
I think the underlying issue is a different treatment of floating point and is 'just' a numerical precision variance.
Perhaps a close to 0 or close to 24 check makes sense here.

**Comment by jklenzing** [comment]
> > The conditions for the slt tests should be updated to use a modulus for 24 hours. The difference at one point is 23.999556885555556.
> 
> Modulus isn't going to help here, less than 24.
> I think the underlying issue is a different treatment of floating point and is 'just' a numerical precision variance.
> Perhaps a close to 0 or close to 24 check makes sense here.

Could just compare the sine of the difference with the appropriate wavelength

**Comment by rstoneback** [comment]
> > > The conditions for the slt tests should be updated to use a modulus for 24 hours. The difference at one point is 23.999556885555556.
> > 
> > 
> > Modulus isn't going to help here, less than 24.
> > I think the underlying issue is a different treatment of floating point and is 'just' a numerical precision variance.
> > Perhaps a close to 0 or close to 24 check makes sense here.
> 
> Could just compare the sine of the difference with the appropriate wavelength

Yeah, I like that. Both sine and cosine though so we don't allow for any other points.

**Comment by jklenzing** [comment]
If the datetimes are wrong in both local time and netcdf, there could be a common problem.

**Comment by rstoneback** [comment]
> If the datetimes are wrong in both local time and netcdf, there could be a common problem.

The first time in the Instrument that generates the netcdf is 2009-01-01 23:58:20. The first time to come back out is 1969-12-31 23:59:57.852516352

My guess is the seconds written to file/or read from aren't getting properly scaled. Not sure why though...

**Comment by rstoneback** [comment]
netcdf works now. Still failing on the SLT test,
```
>       assert np.max(np.abs(sin_diff)) < 1.0e-6
94
E       AssertionError: assert 0.00011600709002063776 < 1e-06
```
which passes on the Mac. What's up with precision on windows?

**Comment by rstoneback** [comment]
My guess is that `float` and `int` default to single precision in windows, but default to doubles on unix systems. Output of current unit tests will tell the tale.

File: pysat/instruments/methods/testing.py [hunk 1]
```diff
@@ -342,9 +342,9 @@ def generate_times(fnames, num, freq='1S'):
     for loop, fname in enumerate(fnames):
         # Grab date from filename
         parts = os.path.split(fname)[-1].split('-')
-        yr = int(parts[0])
-        month = int(parts[1])
-        day = int(parts[2][0:2])
+        yr = np.int(parts[0])
+        month = np.int(parts[1])
+        day = np.int(parts[2][0:2])
```
Comment (rstoneback) [inline pysat/instruments/methods/testing.py:347]:
```suggestion
        yr = int(parts[0])
        month = int(parts[1])
        day = int(parts[2][0:2])
```

File: pysat/instruments/methods/testing.py [hunk 2]
```diff
@@ -299,9 +299,9 @@ def generate_fake_data(t0, num_array, period=5820, data_range=[0.0, 24.0],
     if cyclic:
         uts_root = np.mod(t0, period)
         data = (np.mod(uts_root + num_array, period)
-                * (np.diff(data_range)[0] / float(period))) + data_range[0]
+                * (np.diff(data_range)[0] / np.float(period))) + data_range[0]
     else:
-        data = ((t0 + num_array) / period).astype(int)
+        data = ((t0 + num_array) / period).astype(np.int)
```
Comment (rstoneback) [inline pysat/instruments/methods/testing.py:304]:
```suggestion
        data = ((t0 + num_array) / period).astype(int)
```

File: pysat/instruments/methods/testing.py [hunk 3]
```diff
@@ -299,9 +299,9 @@ def generate_fake_data(t0, num_array, period=5820, data_range=[0.0, 24.0],
     if cyclic:
         uts_root = np.mod(t0, period)
         data = (np.mod(uts_root + num_array, period)
-                * (np.diff(data_range)[0] / float(period))) + data_range[0]
+                * (np.diff(data_range)[0] / np.float(period))) + data_range[0]
```
Comment (rstoneback) [inline pysat/instruments/methods/testing.py:302]:
```suggestion
                * (np.diff(data_range)[0] / np.float64(period))) + data_range[0]
```

File: pysat/utils/coords.py [hunk 4]
```diff
@@ -127,7 +127,7 @@ def calc_solar_local_time(inst, lon_name=None, slt_name='slt',
 
     # Convert from numpy epoch nanoseconds to UT seconds of day
     ut_hr = list()
-    for nptime in inst.index.values.astype(int):
+    for nptime in inst.index.values.astype(np.int):
```
Comment (jklenzing) [inline pysat/utils/coords.py:130]:
```suggestion
    for nptime in inst.index.values.astype(np.int64):
```

**Comment by rstoneback** [comment]
Thanks @jklenzing!

File: pysat/utils/coords.py [hunk 5]
```diff
@@ -145,7 +145,7 @@ def calc_solar_local_time(inst, lon_name=None, slt_name='slt',
     # Calculate solar local time
     if inst[lon_name].shape == ut_hr.shape or inst[lon_name].shape == ():
         if inst[lon_name].shape == ():
-            lon = float(inst[lon_name])
+            lon = np.float(inst[lon_name])
```
Comment (rstoneback) [inline pysat/utils/coords.py:148]:
```suggestion
            lon = np.float64(inst[lon_name])
```

File: pysat/_instrument.py [hunk 6]
```diff
@@ -3425,7 +3425,7 @@ def to_netcdf4(self, fname=None, base_instrument=None, epoch_name='Epoch',
             # attach metadata
             cdfkey.setncatts(new_dict)
 
-            # attach data
+            # Attach Time index data
```
Comment (aburrell) [inline pysat/_instrument.py:3428]:
```suggestion
            # Attach the time index to the data
```

File: pysat/tests/test_utils_coords.py [hunk 7]
```diff
@@ -93,8 +93,14 @@ def test_calc_solar_local_time(self, name):
...
 
         # This works because test instrument longitude ranges from 0-360 deg
-        assert (abs(self.py_inst['slt'].values
-                    - self.py_inst['longitude'].values / 15.0)).max() < 1.0e-6
+        # Testing the difference in periodic space to guard against changes
+        # in numerical precision across platforms.
+        diff = abs(self.py_inst['slt'].values
+                   - self.py_inst['longitude'].values / 15.0)
+        sin_diff = np.sin(diff * 2. * np.pi / 24.)
+        cos_diff = np.cos(diff * 2. * np.pi / 24.)
```
Comment (aburrell) [inline pysat/tests/test_utils_coords.py:101]:
```suggestion
        diff_radians = diff * np.pi / 12.0
        sin_diff = np.sin(diff_radians)
        cos_diff = np.cos(diff_radians)
```

**Review by aburrell — COMMENTED** [review]
Should `set_ga_dir.py` be in the `.github` directory or the `tests` directory?

**Comment by rstoneback** [comment]
> Should `set_ga_dir.py` be in the `.github` directory or the `tests` directory?

I'll leave that one for @jklenzing.

**Comment by rstoneback** [comment]
Latest commits, in-progress during review by @aburrell, are my attempt to proactively seek out `int` and `float` calls. `Meta` is the area I'm most concerned about, particularly the labels.

**Comment by jklenzing** [comment]
> > Should `set_ga_dir.py` be in the `.github` directory or the `tests` directory?
> 
> I'll leave that one for @jklenzing.

I like the idea of moving it, but we would have to navigate directories locally or use if blocks to point to the right directory, e.g.
```
python .github/set_ga_dir.py
```
vs
```
python .github\set_ga_dir.py
```
I would prefer the .github directory, as this will change for different CI setups

**Comment by jklenzing** [comment]
Since the entire point of the script was to be os independent, it doesn't make sense to use the solution above.  See #809 for a cleaner suggestion.  Can't add a code suggestion on deleted lines.

**Comment by jklenzing** [comment]
image

I'm just as guilty of this as anyone, but this come to mind every time I check the Github Actions queue

File: pysat/_meta.py [hunk 8]
```diff
@@ -30,7 +30,8 @@ class Meta(object):
         that have the label values and value types in that order.
         (default={'units': ('units', str), 'name': ('long_name', str),
         'notes': ('notes', str), 'desc': ('desc', str),
-        'min_val': ('value_min', float), 'max_val': ('value_max', float),
+        'min_val': ('value_min', float),
+        'max_val': ('value_max', float),
```
Comment (aburrell) [inline pysat/_meta.py:34]: Why was this broken on two lines?

File: pysat/_meta.py [hunk 9]
```diff
@@ -1270,9 +1271,11 @@ class MetaLabels(object):
     desc : tuple
         Description label name and value type (default=('desc', str))
     min_val : tuple
-        Minimum value label name and value type (default=('value_min', float))
+        Minimum value label name and value type (default=('value_min',
+        float))
     max_val : tuple
-        Maximum value label name and value type (default=('value_max', float))
+        Maximum value label name and value type (default=('value_max',
+        float))
```
Comment (aburrell) [inline pysat/_meta.py:1278]: Again, it looks like there's no reason for this change.

File: pysat/_meta.py [hunk 10]
```diff
@@ -1333,7 +1336,8 @@ class MetaLabels(object):
     def __init__(self, metadata=None, units=('units', str),
                  name=('long_name', str), notes=('notes', str),
                  desc=('desc', str),
-                 min_val=('value_min', float), max_val=('value_max', float),
+                 min_val=('value_min', float),
+                 max_val=('value_max', float),
```
Comment (aburrell) [inline pysat/_meta.py:1340]: Also here.

File: pysat/_meta.py [hunk 11]
```diff
@@ -1465,18 +1469,48 @@ def default_values_from_type(self, val_type):
...
+        # Perform some pre-checks on type, checks that could error with
+        # unexepected input.
+        try:
+            floating_check = isinstance(val_type(), np.floating)
+        except TypeError as err:
+            if str(err).find('not a callable function') > 0:
+                floating_check = False
+            else:
+                # Unexpected input
+                floating_check = None
```
Comment (aburrell) [inline pysat/_meta.py:1485]: I don't think this works.  Example:
```
import numpy as np
val_type = np.asarray([15]).astype(np.float64)[0]
type(val_type)
numpy.float64

floating_check = isinstance(val_type(), np.floating)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-23-4e16dfc35319> in <module>
----> 1 floating_check = isinstance(val_type(), np.floating)

TypeError: 'numpy.float64' object is not callable
```

This yields `floating_check = False`.  Following:
```
val_type is float or floating_check
False
```
Which is not really what you were going for, I believe.

File: pysat/tests/test_meta.py [hunk 12]
```diff
@@ -67,6 +70,43 @@ def test_default_label_value_raises_error(self):
...
+            captured = caplog.text
+
+            # Test for expected string
+            test_str = 'No type match found for '
+            assert captured.find(test_str) >= 0
+
+        return
+
+    @pytest.mark.parametrize("input",
+                             [float, np.float64, np.float32])
```
Comment (aburrell) [inline pysat/tests/test_meta.py:89]:
```suggestion
                             [float, np.float16, np.float32, np.float64, np.float128])
```

File: pysat/_meta.py (hunk 11 above)
Comment (rstoneback) [reply pysat/_meta.py:1490]: `val_type` in your example is a float number. The `default_vals_from_type` function is used when a meta variable is created and looks to the MetaLabels class. It is expecting an input like `float` or `np.float64`.

`labels = pysat._meta.MetaLabels(max_val=('max_val', float), min_val=('min_val', np.float64))`

If a user supplies a number instead then:
```
In [11]: meta = pysat.Meta(labels={'min_val':('min_val', 17)})                                                                                                                                                              

In [12]: meta                                                                                                                                                                                                               
Out[12]: 
pysat.Meta(metadata=Empty DataFrame
Columns: [units, long_name, notes, desc, min_val, value_max, fill]
Index: [], labels=pysat.MetaLabels(units=units <class 'str'>, name=long_name <class 'str'>, notes=notes <class 'str'>, desc=desc <class 'str'>, min_val=min_val 17, max_val=value_max <class 'float'>, fill_val=fill <class 'float'>)export_nan=[]) -> 0 Variables

In [13]: meta['hi'] = {'units': 'time'}                                                                                                                                                                                     
pysat INFO: No type match found for 17

```

File: pysat/tests/test_meta.py (hunk 12 above)
Comment (rstoneback) [reply pysat/tests/test_meta.py]: GitHub Actions doesn't support np.float128 but I've taken the rest of your suggestion.

File: pysat/_meta.py (hunk 8 above)
Comment (rstoneback) [reply pysat/_meta.py]: Fixed.

File: pysat/_meta.py (hunk 9 above)
Comment (rstoneback) [reply pysat/_meta.py]: See above but fixed.

File: pysat/_meta.py (hunk 10 above)
Comment (rstoneback) [reply pysat/_meta.py]: See above but fixed.

File: pysat/_meta.py (hunk 11 above)
Comment (rstoneback) [reply pysat/_meta.py:1490]: Latest commits updates this to:
```
In [1]: import pysat                                                                                                                                                                                                        

In [2]: pysat.logger.setLevel(1)                                                                                                                                                                                            

In [3]: meta = pysat.Meta(labels={'min_val':('min_val', 17)})                                                                                                                                                               

In [4]: meta['hi'] = {'units': 'time'}                                                                                                                                                                                      
pysat INFO: No type match found for 17
pysat INFO: A problem may have been encountered with the user supplied type for Meta attribute:  min_val Please check the settings provided to `labels` at Meta instantiation.
```

Comment (aburrell) [reply pysat/_meta.py:1490]: 🐶  Well that wasn't the smartest by me.  Checked again with the right input type and it works.

**Review by aburrell — APPROVED** [review]

File: pysat/_meta.py (hunk 11 above)
Comment (rstoneback) [reply pysat/_meta.py:1490]: no worries 😄

**Comment by rstoneback** [comment]
Thanks @aburrell for the review and @jklenzing for the code.

**Comment by jklenzing** [comment]
I added the other windows tests issues and closed those out.  Leaving the main one open pending further discussion.

**Pull request merged**